subtype/subtype.go

```diff
@@ -45,27 +48,93 @@ func (s *subtypeSvc) Resource(name string) interface{} {
 	return nil
 }
 
-// Replace replaces all resources with r.
-func (s *subtypeSvc) Replace(r map[resource.Name]interface{}) error {
+// ReplaceAll replaces all resources with r.
+func (s *subtypeSvc) ReplaceAll(r map[resource.Name]interface{}) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 	resources := make(map[string]interface{}, len(r))
 	shortNames := make(map[string]string, len(r))
+	s.resources = resources
+	s.shortNames = shortNames
 	for n, v := range r {
-		if n.Name == "" {
-			return errors.Errorf("Empty name used for resource: %s", n)
+		if err := s.doAdd(n, v); err != nil {
+			return err
 		}
-		name := n.ShortName()
-		resources[name] = v
-		shortcut := name[strings.LastIndexAny(name, ":")+1:]
-		if _, ok := shortNames[shortcut]; ok {
-			shortNames[shortcut] = ""
+	}
+	return nil
+}
+
+func (s *subtypeSvc) Add(n resource.Name, iface interface{}) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.doAdd(n, iface)
+}
+
+func (s *subtypeSvc) Remove(n resource.Name) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.doRemove(n)
+}
+
+func (s *subtypeSvc) ReplaceOne(n resource.Name, iface interface{}) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	err := s.doRemove(n)
+	if err != nil {
+		return err
+	}
+	return s.doAdd(n, iface)
+}
+
+func (s *subtypeSvc) doAdd(n resource.Name, iface interface{}) error {
+	if n.Name == "" {
+		return errors.Errorf("empty name used for resource: %s", n)
+	}
+	name := n.ShortName()
+
+	_, exists := s.resources[name]
+	if exists {
+		return errors.Errorf("resource %s already exists", n)
+	}
+
+	s.resources[name] = iface
+	shortcut := getShortcutName(name)
+	if shortcut != name {
```

Comment:
I think we should remove this check, if both a local resource "NameA" and a remote resource "remote1:NameA" exists the shortcut shouldn't be registered

Reply:
looks good thanks :)